internal/syshandlers/syshandler_tokenpool.go

```diff
@@ -0,0 +1,128 @@
+// Copyright © 2021 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syshandlers
+
+import (
+	"context"
+
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+func (sh *systemHandlers) persistTokenPool(ctx context.Context, pool *fftypes.TokenPoolAnnouncement) (valid bool, err error) {
+	// Find a matching operation within this transaction
+	fb := database.OperationQueryFactory.NewFilter(ctx)
+	filter := fb.And(
+		fb.Eq("tx", pool.TX.ID),
+		fb.Eq("type", fftypes.OpTypeTokensAnnouncePool),
+	)
```

Comment:
Is this query good enough? Could we possibly expect multiple ops of the same type within this transaction..?

Reply:
The more I think about it, I'm very wary of that breaking idempotence. Feels like we'd need a stronger mapping between an operation and its side-effects if we want to support "retries" being equivalent with "adding another operation of the same type within the transaction". Maybe just need to revisit this when we do look at adding retries...

Reply:
>  Maybe just need to revisit this when we do look at adding retries...

Going to raise an architecture tagged issue on this, as it's a V1.0 item to put in the retry support. I agree we need to spend the time to get it right, and this isn't the time/place 👍

Reply:
https://github.com/hyperledger/firefly/issues/221 raised